zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1839,12 +1848,15 @@ protected void doRetry() {
     });
   }
 
-  private void checkDataSizeLimit(byte[] data) {
-    if (data != null && data.length > ZNRecord.SIZE_LIMIT) {
-      LOG.error(
-          "Data size larger than 1M, will not write to zk. Data (first 1k): " + new String(data)
-              .substring(0, 1024));
-      throw new ZkClientException("Data size larger than 1M");
+  private void checkDataSizeLimit(String path, byte[] data) {
+    if (data == null) {
+      return;
+    }
+
+    int sizeLimit = ZNRecordUtil.getSerializerWriteSizeLimit();
+    if (data.length > sizeLimit) {
+      throw new ZkMarshallingError(
```

Comment:
I agree that this is the right thing to do. Thanks for catching this and fixing this.

One concern I have is with the change of the exception type. I am not sure if changing it to ZkMarshallingError is the right thing to do because

1. ZkMarshallingError is usually thrown in ZkSerializer implementations.
2. Changing Exception types may cause backward-compatibility issues.

What do you think?

Reply:
These questions are valid and natural ones which I also thought about.
1. The reason I change it is I would like to make it consistent with serializer. In helix zkclient, the data size check is right after serializing. `ZkMarshallingError` could also indicate the serialized data still exceeds the limit. So I feel `ZkMarshallingError` is more accurate than `ZkClientException`.
2. Actually before version 1.0.0, the exception type is `HelixException`. So changing to `ZkClientException` already breaks backward-compatibility if it breaks. Considering recent 1.0.0 zookeeper-api module users(helix-core 1.0.0 is deprecated), if a user explicitly uses this raw zkclient and catches this runtime exception, that may break. But I am not sure if that really exists in 1.0.0 usage, considering the APIs don't explicitly signature `ZkClientException`.

Though having these thoughts, I don't have a strong preference: we could keep the original exception type. If we would like to change the type, may be it is a good idea to do it in 1.0.1 to avoid future backward compatibility issue.

Reply:
To give you more context, ZkMarshallingError is something that was inherited from 101tec.zkclient library and explicitly thrown for ZkSerializer implementations. Since this is our ZkClient API-level check outside the serializer, I think ZkClientException. 

Also, we've decided to convert all HelixExceptions to ZkClientException, so I think ZkClientException would be more appropriate.

Reply:
+1. Let's not change the exception type of this.

Reply:
Thanks for the input. Resolved.